File: VB/AsynchronousFindInTreeList/AsyncSearchHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Data
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraTreeList
Imports System.ComponentModel
Imports System.Threading
Imports System.Windows.Forms

Namespace AsynchronousFindInTreeList
	Public Class AsyncSearchHelper
		Public Event AsyncOperationIsComplete As EventHandler
		Private _treeList As TreeList
		Private bgwTextFinder As BackgroundWorker
		Private Shared FindText As String

		Public Sub New(ByVal treeList As TreeList)
			_treeList = treeList
			InitializeBackgroundWorker()
		End Sub
		Public Sub AsyncSearch(ByVal table As DataTable, ByVal text As String)
			Dim h As New SearchHelper() With {.Table = table, .FindText = text, .KeyFieldName = _treeList.KeyFieldName, .ParentFieldName = _treeList.ParentFieldName}
			_treeList.Enabled = False
			bgwTextFinder.RunWorkerAsync(h)
		End Sub
		Private Sub InitializeBackgroundWorker()
			bgwTextFinder = New BackgroundWorker()
			AddHandler bgwTextFinder.DoWork, AddressOf bgwTextFinder_DoWork
			AddHandler bgwTextFinder.RunWorkerCompleted, AddressOf bgwTextFinder_RunWorkerCompleted
			bgwTextFinder.WorkerReportsProgress = True
		End Sub
		Private Sub bgwTextFinder_DoWork(ByVal sender As Object, ByVal e As DoWorkEventArgs)
			Dim helper As SearchHelper = TryCast(e.Argument, SearchHelper)
			Dim searchResult As List(Of Integer) = SearchInDataSet(helper.Table, helper.KeyFieldName, helper.ParentFieldName, helper.FindText)
			e.Result = searchResult
		End Sub
		Private Sub bgwTextFinder_RunWorkerCompleted(ByVal sender As Object, ByVal e As RunWorkerCompletedEventArgs)
			Dim res As List(Of Integer) = TryCast(e.Result, List(Of Integer))
			If res Is Nothing Then
				Return
			End If
			Dim iterator As New ExpandingToSpecificNode(res)
			CType(New Thread(New ThreadStart(Function() AnonymousMethod1(iterator, iterator))), Thread).Start()
		End Sub
		
        Private Function AnonymousMethod1(ByVal iterator As ExpandingToSpecificNode, ByVal _iterator As ExpandingToSpecificNode) As Boolean
            _treeList.BeginUpdate()
            Try
                _treeList.NodesIterator.DoOperation(iterator)
            Finally
                _treeList.Invoke(New MethodInvoker(Function() AnonymousMethod2(_iterator)))
            End Try
            Return True
        End Function
		
		Private Function AnonymousMethod2(ByVal iterator As ExpandingToSpecificNode) As Boolean
			_treeList.EndUpdate()
			_treeList.Enabled = True
			Me.RaiseAsyncOperationIsComplete()
			iterator.SelectedNode.Selected = True
			Return True
		End Function
		Private Function SearchInDataSet(ByVal table As DataTable, ByVal keyFieldName As String, ByVal parentFieldName As String, ByVal findText As String) As List(Of Integer)
			If table Is Nothing OrElse keyFieldName Is Nothing OrElse parentFieldName Is Nothing OrElse findText Is Nothing Then
				Return Nothing
			End If
			AsyncSearchHelper.FindText = findText
			Dim result As New List(Of Integer)()
			Dim foundRow As DataRow = table.AsEnumerable().ToList().Find(AddressOf FindRow)
			If foundRow IsNot Nothing Then
				Dim id As Integer = Convert.ToInt32(foundRow.Field(Of Double)(keyFieldName))
				Do While id <> 0
					result.Add(id)
					Dim query As IEnumerable(Of DataRow) = _
						From product In table.AsEnumerable() _
						Where product.Field(Of Double)(keyFieldName) = id _
						Select product
					id = Convert.ToInt32(query.ToArray()(0).Field(Of Double)(parentFieldName))
				Loop
			Else
				Console.WriteLine("Nothing was found...")
				Return Nothing
			End If
			result.Sort()
			Return result
		End Function

		Private Shared Function FindRow(ByVal row As DataRow) As Boolean
			Dim colNames As New List(Of String)()
			For Each item In row.ItemArray
				If item.ToString().Equals(FindText) Then
					Return True
				End If
			Next item
			Return False
		End Function
		Private Sub RaiseAsyncOperationIsComplete()
			RaiseEvent AsyncOperationIsComplete(Me, New EventArgs())
		End Sub
	End Class
End Namespace

File: CS/AsynchronousFindInTreeList/AsyncSearchHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Data;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraTreeList;
using System.ComponentModel;
using System.Threading;
using System.Windows.Forms;

namespace AsynchronousFindInTreeList {
    public class AsyncSearchHelper {
        public event EventHandler AsyncOperationIsComplete;
        TreeList _treeList;
        BackgroundWorker bgwTextFinder;
        static string FindText;

        public AsyncSearchHelper(TreeList treeList) {
            _treeList = treeList;
            InitializeBackgroundWorker();
        }
        public void AsyncSearch(DataTable table, string text) {
            SearchHelper h = new SearchHelper() {
                Table = table,
                FindText = text,
                KeyFieldName = _treeList.KeyFieldName,
                ParentFieldName = _treeList.ParentFieldName
            };
            _treeList.Enabled = false;
            bgwTextFinder.RunWorkerAsync(h);
        }
        void InitializeBackgroundWorker() {
            bgwTextFinder = new BackgroundWorker();
            bgwTextFinder.DoWork += new DoWorkEventHandler(bgwTextFinder_DoWork);
            bgwTextFinder.RunWorkerCompleted += new RunWorkerCompletedEventHandler(bgwTextFinder_RunWorkerCompleted);
            bgwTextFinder.WorkerReportsProgress = true;
        }
        void bgwTextFinder_DoWork(object sender, DoWorkEventArgs e) {
            SearchHelper helper = e.Argument as SearchHelper;
            List<int> searchResult = SearchInDataSet(helper.Table, helper.KeyFieldName, helper.ParentFieldName, helper.FindText);
            e.Result = searchResult;
        }
        void bgwTextFinder_RunWorkerCompleted(object sender, RunWorkerCompletedEventArgs e) {
            List<int> res = e.Result as List<int>;
            if(res == null) return;
            ExpandingToSpecificNode iterator = new ExpandingToSpecificNode(res);
            new Thread(new ThreadStart(() => {
                _treeList.BeginUpdate();
                try {
                    _treeList.NodesIterator.DoOperation(iterator);
                }
                finally {
                    _treeList.Invoke(new MethodInvoker(() => {
                        _treeList.EndUpdate();
                        _treeList.Enabled = true;
                        this.RaiseAsyncOperationIsComplete();
                        iterator.SelectedNode.Selected = true;
                    }));
                }
            }
            )).Start();
        }
        List<int> SearchInDataSet(DataTable table, string keyFieldName, string parentFieldName, string findText) {
            if(table == null || keyFieldName == null || parentFieldName == null || findText == null) return null;
            FindText = findText;
            List<int> result = new List<int>();
            DataRow foundRow = table.AsEnumerable().ToList().Find(FindRow);
            if(foundRow != null) {
                int id = Convert.ToInt32(foundRow.Field<double>(keyFieldName));
                while(id != 0) {
                    result.Add(id);
                    IEnumerable<DataRow> query =
                        from product in table.AsEnumerable()
                        where product.Field<double>(keyFieldName) == id
                        select product;
                    id = Convert.ToInt32(query.ToArray()[0].Field<double>(parentFieldName));
                }
            }
            else {
                Console.WriteLine("Nothing was found...");
                return null;
            }
            result.Sort();
            return result;
        }
        
        static bool FindRow(DataRow row) {
            List<string> colNames = new List<string>();
            foreach(var item in row.ItemArray) {
                if(item.ToString().Equals(FindText))
                    return true;
            }
            return false;
        }
        void RaiseAsyncOperationIsComplete() {
            if(AsyncOperationIsComplete != null) {
                AsyncOperationIsComplete(this, new EventArgs());
            }
        }
    }
}